File: agents/retrieval-control/ccc-scope.ts
// Did anything ccc indexes change between two commits? The freshness gate's watermark records a
// git HEAD, and in a repo where ten agents commit every minute or two HEAD moves constantly —
// mostly through append-only records the index excludes. A HEAD mismatch alone therefore said
// "stale" about indexes that a re-index would reproduce byte for byte (firedancer 2026-09-25:
// NO_INDEX, then `index` failing twice on "HEAD moved", three minutes to one search).
//
// The question is answered by ccc's own matcher, run under ccc's own interpreter (ccc_scope.py),
// never by a reimplementation of its globs. Any failure to answer returns null, which callers
// treat as "in scope" — the old, conservative verdict.

import { readFileSync, realpathSync } from "node:fs";
import { join } from "node:path";

// The `ccc` entry point is a uv-tool script whose shebang names the interpreter that can import
// cocoindex_code.
function cccPython(ccc: string): string | null {
  try {
    const first =
      readFileSync(realpathSync(ccc), "utf8").split("\n", 1)[0] ?? "";
    const interpreter = first.startsWith("#!/") ? first.slice(2).trim() : "";
    // Only a Python interpreter can import cocoindex_code; anything else (a test double, a
    // wrapper script) cannot answer, and running ccc_scope.py under it would be meaningless.
    return /\/python[0-9.]*$/.test(interpreter) ? interpreter : null;
  } catch {
    return null;
  }
}

async function capture(
  cmd: string[],
  cwd: string,
  stdin?: string,
): Promise<string | null> {
  const child = Bun.spawn({
    cmd,
    cwd,
    stdin: stdin === undefined ? "ignore" : new Response(stdin),
    stdout: "pipe",
    stderr: "ignore",
    signal: AbortSignal.timeout(30_000),
  });
  const [out, code] = await Promise.all([
    new Response(child.stdout).text(),
    child.exited,
  ]);
  return code === 0 ? out : null;
}

export type ScopeDrift = { changed: number; inScope: string[] };

// null = could not decide (treat as stale). Otherwise the changed paths and the subset ccc
// would index; an empty `inScope` means a re-index at `to` reproduces the index built at `from`.
export async function inScopeChanges(
  project: string,
  ccc: string,
  from: string,
  to: string,
): Promise<ScopeDrift | null> {
  const diff = await capture(
    ["git", "-C", project, "diff", "--name-only", "-z", from, to],
    project,
  );
  if (diff === null) return null;
  const changed = diff.split("\0").filter((p) => p !== "");
  if (changed.length === 0) return { changed: 0, inScope: [] };
  const python = cccPython(ccc);
  if (python === null) return null;
  const out = await capture(
    [python, join(import.meta.dir, "ccc_scope.py"), project],
    project,
    JSON.stringify(changed),
  );
  if (out === null) return null;
  try {
    const inScope: unknown = JSON.parse(out);
    return Array.isArray(inScope) &&
      inScope.every((p): p is string => typeof p === "string")
      ? { changed: changed.length, inScope }
      : null;
  } catch {
    return null;
  }
}

File: agents/retrieval-control/ccc_scope.py
"""Which of these project-relative paths would ccc index? Run by ccc-scope.ts under ccc's OWN
interpreter (the `ccc` script's shebang), so the answer comes from ccc's matcher itself —
include/exclude globs plus nested .gitignore — not from a reimplementation that could drift.

stdin: JSON list of paths relative to the project root. argv[1]: project root.
stdout: JSON list of the paths that are in scope.

Every approximation errs toward IN scope, because in scope means "the index may be stale":
- the size limit is not applied (a too-large file would only have been dropped);
- a changed .gitignore or .cocoindex_code/settings.yml changes the scope itself, so it counts.
"""

import json
import sys
from pathlib import Path, PurePath

from cocoindex_code.file_walk import build_matcher
from cocoindex_code.settings import load_project_settings


def main() -> None:
    root = Path(sys.argv[1])
    paths = json.load(sys.stdin)
    settings = load_project_settings(root)
    matcher = build_matcher(root, settings.include_patterns, settings.exclude_patterns)

    def in_scope(path: str) -> bool:
        pure = PurePath(path)
        if pure.name == ".gitignore" or path == ".cocoindex_code/settings.yml":
            return True
        # ccc prunes an excluded directory before it ever looks at the files inside.
        for depth in range(1, len(pure.parts)):
            if not matcher.is_dir_included(PurePath(*pure.parts[:depth])):
                return False
        return matcher.is_file_included(pure)

    json.dump([p for p in paths if in_scope(p)], sys.stdout)


main()
